### ISSUE TITLE
Passing context in a reply in a group chat

### ISSUE THREAD

**Issue by robertoash** [n3d1117/chatgpt-telegram-bot#348]
First of all, thank you so much for coding this! Amazing work!

I cannot for the life of me figure out how to the bot to answer taking into consideration reply context in group chats, though. Reading the documentation and some closed issues leads me to believe that it should be possible but I might need some help at this point.

This is what happens when I forward an article link to the bot in a private chat and ask it to summarize it:

![IMG_3799](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/99d2c0f6-d290-46fc-a406-ee682319cb52)

It sort of works. Telegram treats the message and the actual forwarding as 2 different message and that causes the bot to produce two replies. But I guess that would be hard to correct seeing that the bot receives what it thinks is two separate messages.

Anyway, the issue comes when I try to do sort of the same task but with a reply in a group chat. Someone sends a link to the group chat. I reply over that message and write "/chat can you summarize this for me?". But the bot has no context in that case. Here's what happens:

![IMG_3800](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/9d801f4c-7eed-4126-a915-a5c6bd16c100)

I'm a bit unsure about the use of "/chat" because I saw it in an issue here but not in the documentation. Am I doing something wrong? Or are my environment variables wrongly set? Here's the relevant portion of my .env file:
```
# Telegram user ID of admins, or - to assign no admin
ADMIN_USER_IDS=XXXXX

# Comma separated list of telegram user IDs, or * to allow all
ALLOWED_TELEGRAM_USER_IDS=XXXXX,YYYYY,ZZZZZ

# Optional configuration, refer to the README for more details
BUDGET_PERIOD=daily
USER_BUDGETS=0.25,0.25,0.25
GUEST_BUDGET=0
# TOKEN_PRICE=0.002
# IMAGE_PRICES=0.016,0.018,0.02
# TRANSCRIPTION_PRICE=0.006
# ENABLE_QUOTING=true
# ENABLE_IMAGE_GENERATION=true
ENABLE_TRANSCRIPTION=false
# PROXY=http://localhost:8080
OPENAI_MODEL=gpt-3.5-turbo
# ASSISTANT_PROMPT="You are a helpful assistant."
# SHOW_USAGE=false
# STREAM=true
# MAX_TOKENS=1200
# MAX_HISTORY_SIZE=15
# MAX_CONVERSATION_AGE_MINUTES=180
# VOICE_REPLY_WITH_TRANSCRIPT_ONLY=true
# VOICE_REPLY_PROMPTS="Hi bot;Hey bot;Hi chat;Hey chat"
# N_CHOICES=1
# TEMPERATURE=1.0
# PRESENCE_PENALTY=0.0
# FREQUENCY_PENALTY=0.0
# IMAGE_SIZE=512x512
# GROUP_TRIGGER_KEYWORD=""
# IGNORE_GROUP_TRANSCRIPTIONS=true
# BOT_LANGUAGE=en
```

Any help would be appreciated. Thank you so much!

**Comment by stanislavlysenko0912** [comment]
1. The function of context support in response to the message was implemented by me, and most likely it works only by keyword "GROUP_TRIGGER_KEYWORD", which you do not have

I did not think that this command "/chat" someone uses in the process of communicating, and I'll fix this bug

2. As far as I know, there is no function to view page URLs yet. 
Since there is a process of adding "plugins or functions (whichever is easier)", I think when the basis for them will be ready, someone will do the implementation of such, but there are also nuances in my opinion (the amount of information on the page).
In any case, so the question you need to just wait.

**Comment by robertoash** [comment]
> 1. The function of context support in response to the message was implemented by me, and most likely it works only by keyword "GROUP_TRIGGER_KEYWORD", which you do not have
> 
> I did not think that this command "/chat" someone uses in the process of communicating, and I'll fix this bug
> 

I have also tried adding a GROUP_TRIGGER_KEYWORD. For example, if I set it to "chatbot" then the bot doesn't reply to `/chat hello` but it replies to `/chat chatbot hello`.

I understand the bot has no way of knowing if he is being talked to in a group chat except with some sort of keyword. I think `GROUP_TRIGGER_KEYWORD` should modify the default `/chat` and not add on top of it.

> 2. As far as I know, there is no function to view page URLs yet. 
> 
> Since there is a process of adding "plugins or functions (whichever is easier)", I think when the basis for them will be ready, someone will do the implementation of such, but there are also nuances in my opinion (the amount of information on the page).
> 
> In any case, so the question you need to just wait.

Regarding context, I think the issue is that I can't get the bot (don't know if mine or all) to receive the context at all. As you can see on the first screenshot, I do get a summary of the link in a private conversation. If the link was the issue then it would reply, in both private and group chats, that it can't see the contents of the link I sent and not that it needs me to send a link.

**Comment by stanislavlysenko0912** [comment]
> > 1. Функцию поддержки контекста в ответ на сообщение реализовал я, и скорее всего она работает только по ключевому слову "GROUP_TRIGGER_KEYWORD", которого у вас нет
> > 
> > Не думал, что эту команду "/chat" кто-то использует в процессе общения, и я исправлю этот баг
> 
> Я также попытался добавить GROUP_TRIGGER_KEYWORD. Например, если я установил «чат-бот», то бот не отвечает, `/chat hello`но отвечает на `/chat chatbot hello`.
> 
> Я понимаю, что бот не может узнать, разговаривают ли с ним в групповом чате, кроме как по какому-то ключевому слову. Я думаю `GROUP_TRIGGER_KEYWORD`, следует изменить значение по умолчанию `/chat`, а не добавлять поверх него.
> 
> > 2. Насколько я знаю, пока нет функции просмотра URL-адресов страниц.
> > 
> > Так как идет процесс добавления "плагинов или функций (что проще)", то думаю когда будет готова база для них кто-то займется реализацией таковой, но есть и нюансы на мой взгляд (количество информации на странице).
> > В любом случае, поэтому вопрос нужно просто подождать.
> 
> Что касается контекста, я думаю, проблема в том, что я вообще не могу заставить бота получать контекст. Как вы можете видеть на первом снимке экрана, в приватной беседе все работает, как и ожидалось, и я получаю сводку содержимого ссылки. Если бы проблема заключалась в ссылке, то он ответил бы как в приватных, так и в групповых чатах, что он не может видеть содержимое отправленной мной ссылки, а не что ему нужно, чтобы я отправил ссылку.

Can you show me how you discounted the link and what you wrote in private messages that he was able to summarize it?

**Comment by robertoash** [comment]
> Can you show me how you discounted the link and what you wrote in private messages that he was able to summarize it?

I just forwarded the link to the private chat like this:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/4286504c-9ee1-422a-adde-fd18748dffed)

This is the chat:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/db5d681f-9380-4809-91ef-d10f91583865)

And this is what the bot replied:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/5afb80f4-b344-4578-911a-482c4f81b2fd)

I thought it might be that the article I was using to test was written in 2018 and that's why it worked but I just sent it this one written in 2022 and it worked as well: https://www.feednavigator.com/Article/2022/03/31/Study-Krill-meal-shows-potential-as-functional-ingredient-in-shrimp-feed

![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/56935032-e834-46a8-a985-836f329e3d14)

**Comment by stanislavlysenko0912** [comment]
> I thought it might be that the article I was using to test was written in 2018 and that's why it worked but I just sent it this one written in 2022 and it worked as well: https://www.feednavigator.com/Article/2022/03/31/Study-Krill-meal-shows-potential-as-functional-ingredient-in-shrimp-feed

I still don't think it goes straight through the articles.
The response speed is questionable, it starts responding too quickly.

I tried using the latest version of GPT-4 and it responded about the article in text, which before 3.5 couldn't answer anything. But when I send him other articles, he won't say anything.

But also here is a screenshot where I asked 3.5 several times about the article, it said several times that it could not open it, the translation "Unfortunately, I can not open this link to view the content of the article"

But at some point, he recounted it. This looks very interesting to me )


*Sorry it's not in English, trust me ) He didn't know anything about it at first then he wrote a summary of its contents 😁*

<img width="320" alt="image" src="https://github.com/n3d1117/chatgpt-telegram-bot/assets/86171768/52a26c85-6c2d-4247-b80d-d7e5d9538792">

**Comment by robertoash** [comment]
Could be interesting to investigate this. However, that feels more like an investigation of what the API does with links. 

For this issue, I think this part is the one to solve, if at all possible:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/dbdf3ea6-d567-48fc-8b29-1ef540b7a039)

Feels like this flow should work but doesn't:
- User sends link to group chat
- Someone replies to that using "@Bot" or "/chat" or "GROUP_TRIGGER_KEYWORD"
- Bot should be able to see the context that was replied to, even when not replying directly to the bot

**Comment by robertoash** [comment]
Here's a test not involving a link but just a plain message:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/98cdf269-4be0-4def-a362-9275823c378d)

Bot doesn't have context when tagged in a reply to someone else's message.

Direct interactions work fine:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/42000824/97d8c96d-8830-4a0c-9089-e648dbc7e9ef)

**Comment by stanislavlysenko0912** [comment]
If you want functionality like in private messages, that the bot would understand messages without a reply to them - it's a bad idea.
You'll just waste tokens all the time :)

**Comment by robertoash** [comment]
> If you want functionality like in private messages, that the bot would understand messages without a reply to them - it's a bad idea.

No. I didn't mean it that way. I mean that when you reply to someone else's message in the chat but use "/chat" in your reply then the bot should have access to the message that you replied to.

**Comment by n3d1117** [comment]
Fixed by @stanislavlysenko0912 in #351
